File: 0016-3sum-closest/0016-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        min_diff = float('inf')
        n = len(nums)
        for i in range(n - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            j, k = i + 1, n - 1
            while j < k: 
                curr_sum = nums[i] + nums[j] + nums[k]
                if abs(curr_sum - target) < min_diff:
                    min_diff = abs(curr_sum - target)
                    closest_sum = curr_sum
                if curr_sum > target:
                    k -= 1
                elif curr_sum < target:
                    j += 1
                elif curr_sum == target:
                    return curr_sum 
        return closest_sum

File: 0016-3sum-closest/0016-3sum-closest.go
import (
	"sort"
)

func threeSumClosest(nums []int, target int) int {
	sort.Ints(nums)
	n := len(nums)

	// 1. Initialize result with the first possible triplet
	closestSum := nums[0] + nums[1] + nums[2]

	for i := 0; i < n-2; i++ {
		// Optional: Skip duplicates for 'i' to speed up
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		left, right := i+1, n-1

		for left < right {
			currSum := nums[i] + nums[left] + nums[right]

			// 2. Exact match check (distance is 0)
			if currSum == target {
				return currSum
			}

			// 3. Update closestSum if the new difference is smaller
			if abs(target-currSum) < abs(target-closestSum) {
				closestSum = currSum
			}

			// 4. Move pointers based on comparison with target
			if currSum < target {
				left++
			} else {
				right--
			}
		}
	}
	return closestSum
}

// Helper function for integer absolute value
func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}